File: 04.05.合法二叉搜索树/04.05.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root, LONG_MIN, LONG_MAX);
    }

    bool helper(TreeNode* root, long lower, long upper) {
        if (root == nullptr) {
            return true;
        }
        if (root->val <= lower || root->val >= upper) {
            return false;
        }
        return helper(root->left, lower, root->val) && helper(root->right, root->val, upper);
    }
};


// 中序遍历
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        stack<TreeNode*> st;
        long inorder = LONG_MIN;

        while (!st.empty() || root != nullptr) {
            while (root != nullptr) {
                st.push(root);
                root = root->left;
            }

            root = st.top();
            st.pop();
            if (root->val <= inorder) {
                return false;
            }

            inorder = root->val;
            root = root->right;
        }

        return true;
    }
};

File: 04.05.合法二叉搜索树/04.05.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isValidBST(root *TreeNode) bool {
	return helper(root, math.MinInt64, math.MaxInt64)
}

func helper(root *TreeNode, lower, upper int64) bool {
	if root == nil {
		return true
	}

	if int64(root.Val) <= lower || int64(root.Val) >= upper {
		return false
	}

	return helper(root.Left, lower, int64(root.Val)) && helper(root.Right, int64(root.Val), upper)
}

// 中序遍历
func isValidBST(root *TreeNode) bool {
	st := make([]*TreeNode, 0)
	inorder := math.MinInt64

	for len(st) > 0 || root != nil {
		for root != nil {
			st = append(st, root)
			root = root.Left
		}

		root = st[len(st)-1]
		st = st[:len(st)-1]

		if root.Val <= inorder {
			return false
		}

		inorder = root.Val
		root = root.Right
	}

	return true
}